### PR TITLE
[FLINK-15021][network] Remove the setting of netty channel watermark in NettyServer

### PR DESCRIPTION
**Pull request by zhijiangW**
## What is the purpose of the change

*The high and low watermark setting in NetterServer before was mainly used for network flow control and limiting the maximum memory overhead caused by copying data inside netty stack. In detail, when the downstream side processes slowly and exhausted the available buffers finally, it would temporarily close the auto read switch in netty stack. Then the upstream side would finally reach the high watermark of channel to become unwritable.*

*But based on credit-based flow control and reusing flink network buffer inside netty stack, the watermark setting is not invalid now. So we can safely remove it to cleanup the codes.*

## Brief change log

  - *Remove the watermark setting in `NettyServer`*
  - *Remove unused method from `NettyConfig` class*
  - *Fix the import format of `NettyConfig` class*

## Verifying this change

Cleanup codes without tests coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java [hunk 1]
```diff
@@ -185,13 +185,8 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object msg) throws Exc
 		}
 	}
 
-	@Override
-	public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {
-		writeAndFlushNextMessageIfPossible(ctx.channel());
-	}
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java]: Why don't we need this anymore? Could you explain this a little bit more?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyServer.java [hunk 2]
```diff
@@ -127,20 +127,6 @@ int init(
...
-		final int newLowWaterMark = config.getMemorySegmentSize() + 1;
-		final int newHighWaterMark = 2 * config.getMemorySegmentSize();
-		if (newLowWaterMark > defaultHighWaterMark) {
-			bootstrap.childOption(ChannelOption.WRITE_BUFFER_HIGH_WATER_MARK, newHighWaterMark);
-			bootstrap.childOption(ChannelOption.WRITE_BUFFER_LOW_WATER_MARK, newLowWaterMark);
-		} else { // including (newHighWaterMark < defaultLowWaterMark)
-			bootstrap.childOption(ChannelOption.WRITE_BUFFER_LOW_WATER_MARK, newLowWaterMark);
-			bootstrap.childOption(ChannelOption.WRITE_BUFFER_HIGH_WATER_MARK, newHighWaterMark);
-		}
-
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyServer.java]: Same about this. Could you explain why don't we need this aynmore?

**Comment by zhijiangW** [comment]
Thanks for the review @pnowojski .

I would explain a bit for above two issues. The watermark setting related with writability changed are both for previous network flow control.

In the past, when the downstream side has no available buffers to hold the data, then it would set the `autoRead` as false for the network channel. As a result, the tcp server side would not transport the data to the client side eventually. Finally the server socket cache is full to block the netty flush operation. That means the partition buffer would write into the netty stack but can not flush into the socket side. Also note that we did not reuse the flink buffer in netty stack before, so it would consume more netty internal `ByteBuf` once they can not be flushed immediately. So we set the netty watermark to control how many buffers can be written into netty stack. If the high watermark is reached, then we would not write partition buffer any more. Therefore the partition buffer would not be recycled to block the task thread called back pressure.

As for credit-based flow control, the netty server writes partition buffer into network stack based on credit feedback, so the data would not be accumulated in the socket stack. Moreover the netty server would reuse the flink buffer to not consume internal `ByteBuf` any more. So the netty channel should alway be writable now, then we can remove these invalid logics.

**Comment by zhijiangW** [comment]
I guess we also need to adjust the doc which refers to watermark in some section.
https://cwiki.apache.org/confluence/display/FLINK/Netty+memory+allocation

**Comment by pnowojski** [comment]
Thanks for the explanation. Partially I get it. I'm fine with removing code setting the watermarks. 

However I'm not entirely sure about the `channelWritabilityChanged`. Wouldn't getting rid of this code result in network stack blocking on full system network buffers? True, we have reserved memory on the receiver side, but that doesn't mean we will not overflow the sender buffers (basically when network is the bottleneck). In that case, if we do not check `!channel.isWritable()`, won't we block the Netty's thread?

Regarding the [wiki](65e95074e55c911ddec0053b41402dd923e9b2bc) I would mark it as obsolete and maybe point to some of Nico's [blog post](https://flink.apache.org/2019/06/05/flink-network-stack.html#credit-based-flow-control)?

**Comment by zhijiangW** [comment]
Thanks for the reminder and your concerns made me think a bit more. 

If I remembered correctly, the written messages would be added into the pending entry in netty's `ChannelOutboundBuffer`, so the netty thread would never be blocked in the case of network bottleneck. But there is another memory overhead concern in that case. 

On sender side, we already reused the flink network buffer in netty stack to avoid allocating more chunks for holding the `BufferResponse` message. But we still need to allocate netty internal memory for the header(30 bytes size) of each `BufferResponse`. Therefore if we ignore the high watermark of netty channel and write the `BufferResponse` always, the header memory overhead still can not be controlled properly which might cause direct OOM issue in extreme case.

Considering above concerns, I think it is safe to retain the logics of `channel.isWritable()` and `channelWritabilityChanged`, and only remove the setting of watermark. The previous watermark setting was mainly for controlling the netty chunk overhead when copying the `BufferResponse`, but now the chunk is only used for holding the header and the default watermark 32kb - 64kb is already suitable for us.

>Regarding the wiki I would mark it as obsolete and maybe point to some of Nico's blog post?
Agree

**Comment by pnowojski** [comment]
> If I remembered correctly, the written messages would be added into the pending entry in netty's ChannelOutboundBuffer, so the netty thread would never be blocked in the case of network bottleneck. But there is another memory overhead concern in that case.

Ok, I've looked around and what I've found confirms this.

Regarding the 30 bytes overhead per message, I haven't thought this through. If you think it's better to keep the `channelWritabilityChanged` code, let's do it, as it's easier and doesn't requires much thinking from our side :) I guess with only a couple of records per buffer (like 100 output channels 1 ms output flusher or 1000 channels and 10 ms output flusher), this overhead might be quite significant?

**Comment by zhijiangW** [comment]
> I guess with only a couple of records per buffer (like 100 output channels 1 ms output flusher or 1000 channels and 10 ms output flusher), this overhead might be quite significant?

Exactly, it is up to what is the real data size in message compared with header size. Considering the default high watermark 64KB, if the netty channel is unwritable because of socket cache full caused by network bottleneck, there would be only 60 bytes overhead among 64KB in best case. But in worst case (very small data size in buffer) the overhead might reach 64KB at most.

Actually in above best case we can write more buffers to the netty stack for pending, or in other words we can reach the high watermark, because the total header overhead is really small. It might exist some fine-grained improvements here. I guess we might consider supplementing the micro-benchmark for the case of network bottleneck.

**Comment by pnowojski** [comment]
> It might exist some fine-grained improvements here. I guess we might consider supplementing the micro-benchmark for the case of network bottleneck.

Probably the easiest way to trigger this use case is to set up a single sender thread, with couple of receiver threads? But I'm not sure if this is very high priority, unless it's a low hanging fruit fix or if you think it's an important fix.

**Comment by zhijiangW** [comment]
>Probably the easiest way to trigger this use case is to set up a single sender thread, with couple of receiver threads? 

That may make sense to do so.

>But I'm not sure if this is very high priority, unless it's a low hanging fruit fix or if you think it's an important fix.

It is actually very low priority on my side, just thought of them during the discussion. I would let my team member try to construct this scenario if have free time. :)

**Comment by zhijiangW** [comment]
Thanks for the review @pnowojski !
I updated the codes to revert the deletion of writable changed and only remove the setting of netty watermark as we confirmed before.

**Pull request merged**